core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeUtil.java

```diff
@@ -93,60 +100,82 @@ public static long writePages(PagesSerde serde, SliceOutput sliceOutput, Iterato
         return size;
     }
 
-    public static Iterator<Page> readPages(PagesSerde serde, SliceInput sliceInput)
+    public static Iterator<Page> readPages(PagesSerde serde, InputStream inputStream)
     {
-        return new PageReader(serde, sliceInput);
+        return new PageReader(serde, inputStream);
     }
 
     private static class PageReader
             extends AbstractIterator<Page>
     {
         private final PagesSerde serde;
         private final PagesSerde.PagesSerdeContext context;
-        private final SliceInput input;
+        private final InputStream inputStream;
+        private final byte[] headerBuffer = new byte[SERIALIZED_PAGE_HEADER_SIZE];
+        private final Slice headerSlice = Slices.wrappedBuffer(headerBuffer);
 
-        PageReader(PagesSerde serde, SliceInput input)
+        PageReader(PagesSerde serde, InputStream inputStream)
         {
             this.serde = requireNonNull(serde, "serde is null");
-            this.input = requireNonNull(input, "input is null");
+            this.inputStream = requireNonNull(inputStream, "inputStream is null");
             this.context = serde.newContext();
         }
 
         @Override
         protected Page computeNext()
         {
-            if (!input.isReadable()) {
-                context.close(); // Release context buffers
-                return endOfData();
+            try {
+                int read = ByteStreams.read(inputStream, headerBuffer, 0, headerBuffer.length);
+                if (read <= 0) {
```

Comment:
Is 0 a valid EOF signal here? I thought this would have to be < 0.

Reply:
ByteStreams.read returns the number of bytes read. Won't return a negative number.
<img width="719" alt="image" src="https://user-images.githubusercontent.com/9404831/157718613-3e150174-eb30-4cec-9a55-934200f23f4f.png">

Reply:
Yep, you are correct- makes sense.